### PR TITLE
initial ast and parser structure skeleton

### PR DESCRIPTION
**Pull request by edg-l**
The ast was inspired from our austral port.

The parser in this pr is just enough to parse the factorial program at the bottom.

Current test produces the following output:

Program:

```rust
mod ModuleName {
    const MY_CONSTANT: u8 = 2;
    const MY_CONSTANT2: bool = true;
    const MY_CONSTANT3: string = "hello world!";
}
```

Ast:

```rust
[crates/concrete_parser/src/lib.rs:26] ast = [
    ModuleBase {
        doc_string: None,
        imports: [],
        name: Ident {
            name: "ModuleName",
            span: Span {
                from: 5,
                to: 15,
            },
        },
        contents: [
            Constant(
                ConstantDef {
                    decl: ConstantDecl {
                        doc_string: None,
                        name: Ident {
                            name: "MY_CONSTANT",
                            span: Span {
                                from: 28,
                                to: 39,
                            },
                        },
                        type: Simple {
                            name: Ident {
                                name: "u8",
                                span: Span {
                                    from: 41,
                                    to: 43,
                                },
                            },
                        },
                    },
                    value: Atomic(
                        ConstInt(
                            2,
                        ),
                    ),
                },
            ),
            Constant(
                ConstantDef {
                    decl: ConstantDecl {
                        doc_string: None,
                        name: Ident {
                            name: "MY_CONSTANT2",
                            span: Span {
                                from: 59,
                                to: 71,
                            },
                        },
                        type: Simple {
                            name: Ident {
                                name: "bool",
                                span: Span {
                                    from: 73,
                                    to: 77,
                                },
                            },
                        },
                    },
                    value: Atomic(
                        ConstBool(
                            true,
                        ),
                    ),
                },
            ),
            Constant(
                ConstantDef {
                    decl: ConstantDecl {
                        doc_string: None,
                        name: Ident {
                            name: "MY_CONSTANT3",
                            span: Span {
                                from: 96,
                                to: 108,
                            },
                        },
                        type: Simple {
                            name: Ident {
                                name: "string",
                                span: Span {
                                    from: 110,
                                    to: 116,
                                },
                            },
                        },
                    },
                    value: Atomic(
                        ConstStr(
                            "\"hello world!\"",
                        ),
                    ),
                },
            ),
        ],
    },
]
```

factorial:

```rust
mod FactorialModule {
    pub fn factorial(x: u64) -> u64  {
        return match x {
            0 -> return 1,
            n -> return n * factorial(n-1),
        };
    }
}
```
https://gist.github.com/edg-l/068898bb319fdb324beca30b08c7caf4

### CODE REVIEW

**Review by igaray — APPROVED** [review]

**Review by juanbono — APPROVED** [review]
LGTM

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [workspace]
 resolver = "2"
 
-members = [ "crates/concrete","crates/concrete_ast", "crates/concrete_codegen_mlir", "crates/concrete_driver", "crates/concrete_parser"]
+members = [ "crates/concrete","crates/concrete_ast", "crates/concrete_codegen_mlir", "crates/concrete_driver", "crates/concrete_parser", "crates/concrete_type_checker"]
```
Comment (juanbono) [inline Cargo.toml:4]: what do you think about dropping the `concrete` prefix from the crates? and having:
- ast
- codegen_mlir
- driver
- parser
- type_checker

File: crates/concrete_ast/src/expressions.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+use crate::{operations::Operation, statements::Statement};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Expression {
+    Match(MatchExpr),
+    Operation(Operation),
+    // Block(Vec<Statement>),
```
Comment (unbalancedparentheses) [inline crates/concrete_ast/src/expressions.rs:7]: Shouldn't be commented

**Comment by unbalancedparentheses** [comment]
@igaray mentioned a few things were off, let's fix it in the next pr.

**Pull request merged**